beta-operator-rewards/rewards-requirements.ts

```diff
@@ -0,0 +1,478 @@
+    // This is used to fetch the authorization that was allocated during the rewards
+    // interval.
+    const postIntervalEvents = await filterEventsByStakingProvider(
+      postEvents,
+      stakingProvider
+    );
+
+    return await authorizationPostRewardsInterval(
+      postIntervalEvents,
+      application,
+      stakingProvider
+    );
+  }
+
+  // There is at least one event emitted during the rewards interval
+  const applicationEvents = intervalEvents.filter((obj) => {
+    return obj.args.application == application.address;
+  });
+
+  return await authorizationForRewardsInterval(
+    applicationEvents,
+    startRewardsBlock,
+    endRewardsBlock
+  );
+}
+
+async function filterEventsByApplications(events: any[]) {
+  const filteredEvents = new Array();
+  for (let j = 0; j < events.length; j++) {
+    const event = events[j];
+    if (
+      event.args.application == RandomBeaconAddress ||
+      event.args.application == WalletRegistryAddress
+    ) {
+      filteredEvents.push(event);
+    }
+  }
+
+  return filteredEvents;
+}
+
+async function filterEventsByStakingProvider(
+  events: any[],
+  stakingProvider: string
+) {
+  const filteredEvents = new Array();
+  for (let j = 0; j < events.length; j++) {
+    const event = events[j];
+    if (event.args.stakingProvider == stakingProvider) {
+      filteredEvents.push(event);
+    }
+  }
+
+  return filteredEvents;
+}
+
+// Calculates the weighted authorization for rewards interval based on events.
+// The general idea of weighted rewards calculation describes the following example.
+// Please note that this example operates on dates for simplicity purposes,
+// however the actual calculation is based on block numbers.
+// Ex.
+// events:         ^     ^      *    *   ^
+// timeline:  -|---|-----|------|----|---|--------|-->
+//         Sep 0   3     8      14  18   22       30
+// where: '^' denotes increase in authorization
+//        '*' denotes decrease in authorization
+//         0 -> Sep 1 at 00:00:00
+// event authorizations:
+//  Sep 0 - 3 from 100k to 110k
+//  Sep 3 - 8 from 110k to 135k
+//  Sep 8 - 14 from 135k to 120k
+//  Sep 14 - 18 from 120k to 100k
+//  Sep 18 - 30 constant 100k (last sub-interval)
+// authorization = (3-0)/30*100 + (8-3)/30*110 + (14-8)/30*135
+//               + (18-14)/30*120 + (30-18)/30*100
+async function authorizationForRewardsInterval(
+  events: any[],
+  startRewardsBlock: number,
+  endRewardsBlock: number
+) {
+  if (events.length == 0) {
+    return BigNumber.from("0");
+  }
+
+  let authorization = BigNumber.from("0");
+  const deltaRewardsBlock = endRewardsBlock - startRewardsBlock;
+  // ascending order
+  events.sort(function (a, b) {
+    return a.blockNumber < b.blockNumber
+      ? -1
+      : a.blockNumber == b.blockNumber
+      ? 0
+      : 1;
+  });
+
+  let tmpBlock = startRewardsBlock; // prev tmp block
+  let event;
+  for (let i = 0; i < events.length; i++) {
+    event = events[i];
```

Comment:
I've never actually checked to see what the javascript virtual machine does here, though now I want to know!

Is 
```js
let foo;
for (...) {
  foo = bar()
}
```
more efficient than
```
for (...) {
  let foo = bar()
}
```

It would be in really old languages since we have to allocate `foo` in each loop, but I don't know if the compiler makes the optimization for us in JS or not. I found https://stackoverflow.com/questions/50807989/es6-javascript-const-inside-or-let-outside-loop which is inconclusive.

For the sake of style! I vote for:

```js
for (let i = 0; i < events.length; i++) {
  const event = events[i];
}
```
sine we're not changing the value of event in the loop or using it outside the loop

Reply:
I've actually moved `event` outside the loop to reference it while calculating the last sub-interval authorization `event.arg.amountTo`, but I agree, it can be written nicer. I've just referenced the last event as `events[events.length - 1].args.toAmount.` to calculate the `authorization`. As part of [4505ecf](https://github.com/keep-network/keep-core/pull/3346/commits/4505ecf04eda7051cc0ddea2e336107d3c292fd2)